posthog/hogql/printer.py

```diff
@@ -1,196 +1,298 @@
+
+    # modify the cloned tree as needed
+    if dialect == "clickhouse":
+        # TODO: add team_id checks (currently done in the printer)
+        # TODO: add joins to person and group tables
+        pass
+
+    return _Printer(context=context, dialect=dialect, stack=stack or []).visit(node)
+
+
+@dataclass
+class JoinExprResponse:
+    printed_sql: str
+    where: Optional[ast.Expr] = None
+
 
+class _Printer(Visitor):
+    # NOTE: Call "print_ast()", not this class directly.
+
+    def __init__(
+        self, context: HogQLContext, dialect: Literal["hogql", "clickhouse"], stack: Optional[List[ast.AST]] = None
+    ):
+        self.context = context
+        self.dialect = dialect
+        # Keep track of all traversed nodes.
+        self.stack: List[ast.AST] = stack or []
+
+    def visit(self, node: ast.AST):
+        self.stack.append(node)
+        response = super().visit(node)
+        self.stack.pop()
+        return response
+
+    def visit_select_query(self, node: ast.SelectQuery):
+        if self.dialect == "clickhouse" and not self.context.select_team_id:
+            raise ValueError("Full SELECT queries are disabled if context.select_team_id is not set")
+
+        # We will add extra clauses onto this from the joined tables
         where = node.where
-        # Guard with team_id if selecting from a table and printing ClickHouse SQL
-        # We do this in the printer, and not in a separate step, to be really sure this gets added.
-        # This will be improved when we add proper table and column alias support. For now, let's just be safe.
-        if dialect == "clickhouse" and from_table is not None:
-            where = guard_where_team_id(where, context)
-        where = print_ast(where, stack, context, dialect) if where else None
-
-        having = print_ast(node.having, stack, context, dialect) if node.having else None
-        prewhere = print_ast(node.prewhere, stack, context, dialect) if node.prewhere else None
-        group_by = [print_ast(column, stack, context, dialect) for column in node.group_by] if node.group_by else None
-        order_by = [print_ast(column, stack, context, dialect) for column in node.order_by] if node.order_by else None
 
-        limit = node.limit
-        if context.limit_top_select:
-            if limit is not None:
-                if isinstance(limit, ast.Constant) and isinstance(limit.value, int):
-                    limit.value = min(limit.value, MAX_SELECT_RETURNED_ROWS)
+        select_from = []
+        next_join = node.select_from
+        while isinstance(next_join, ast.JoinExpr):
+            if next_join.symbol is None:
+                raise ValueError("Printing queries with a FROM clause is not permitted before symbol resolution")
+
+            visited_join = self.visit_join_expr(next_join)
+            select_from.append(visited_join.printed_sql)
+
+            # This is an expression we must add to the SELECT's WHERE clause to limit results.
```

Comment:
Took me a while to figure out this was the team ID guard handling 😓 
```suggestion
            # This is an expression we must add to the SELECT's WHERE clause to limit results, like the team ID guard.
```

Reply:
Makes sense for this to be on every table